pulser/simulation/simresults.py

```diff
@@ -54,6 +59,202 @@ def __init__(self, run_output, dim, size, basis_name, meas_basis=None):
+        r"""Returns the result of multiple measurements in a given basis.
+
+        The enconding of the results depends on the meaurement basis. Namely:
+
+        - *ground-rydberg* : :math:`1 = |r\rangle;~ 0 = |g\rangle, |h\rangle`
+        - *digital* : :math:`1 = |h\rangle;~ 0 = |g\rangle, |r\rangle`
+
+        Note:
+            The results are presented using a big-endian representation,
+            according to the pre-established qubit ordering in the register.
+            This means that, when sampling a register with qubits ('q0','q1',
+            ...), in this order, the corresponding value, in binary, will be
+            0Bb0b1..., where b0 is the outcome of measuring 'q0', 'b1' of
+            measuring 'q1' and so on.
+
+        Keyword Args:
+            meas_basis (str, default=None): 'ground-rydberg' or 'digital'. If
+                left as None, uses the measurement basis defined in the
+                original sequence.
+            N_samples (int, default=1000): Number of samples to take.
+
+        Raises:
+            ValueError: If trying to sample without a defined 'meas_basis' in
+                the arguments when the original sequence is not measured.
+        """
+        pass
+
+    @abstractmethod
+    def sample_final_state(self, meas_basis='ground-rydberg', N_samples=1000):
+        pass
+
+
+class NoisyResults(SimulationResults):
+    """
+    Results of a noisy simulation run of a pulse sequence. Contrary to a
+    CleanResults object, this object contains a unique Counter describing the
+    state distribution at the time it was created by using
+    sim.run(spam=True, t).
+
+    Contains methods for studying the populations and extracting useful
+    information from them.
+    """
+
+    def __init__(self, run_output, size, basis_name,
+                 meas_basis="ground-rydberg", dim=2):
+        """
+        Initializes a new NoisyResults instance.
+
+        Warning : Can't have single-atom Hilbert spaces with dimension bigger
+            than 2 for NoisyResults objects.
+            This is not the case for a CleanResults object, containing states
+            in Hilbert space, but NoisyResults contains a probability
+            distribution of bitstrings, not atomic states
+
+        Args:
+            run_output (Counter) : Counter returning the probability of each
+                multi-qubits state, represented as a bitstring.
+            dim (int): The dimension of the local space of each atom (2 or 3).
+            size (int): The number of atoms in the register.
+            basis_name (str): The basis indicating the addressed atoms after
+                the pulse sequence ('ground-rydberg', 'digital' or 'all').
+
+        Keyword Args:
+            meas_basis (None or str): The basis in which a sampling measurement
+                is desired.
+        """
+        super().__init__(run_output, dim, size, basis_name, meas_basis)
+
+    def get_final_state(self):
```

Comment:
Not starting a review, but I just saw this. I don't think it is possible to infer the final density matrix from projective measurements in the z- basis.